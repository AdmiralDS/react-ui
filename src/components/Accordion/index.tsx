import type { FC, HTMLAttributes, KeyboardEvent } from 'react';
import { useCallback, useRef } from 'react';

import { keyboardKey } from '../common/keyboardKey';
import { moveFocus, nextItem, previousItem } from './utils';
import { AccordionWrapper } from './style';

type Dimension = 'l' | 'm';

export * from './AccordionItem';

export interface AccordionProps extends HTMLAttributes<HTMLDivElement> {
  /** Размер компонента */
  dimension?: Dimension;
  /** Расположение иконки шеврона в заголовке. По умолчанию иконка выравнивается по правому краю. */
  iconPosition?: 'right' | 'left';
  /** Скрыть верхний разделитель аккордиона (верхнюю серую полоску).
   * Параметр hideTopDivider скрывает верхнюю серую полоску только у первого AccordionItem, заданного внутри Accordion
   */
  hideTopDivider?: boolean;
  /** Скрыть нижний разделитель аккордиона (нижнюю серую полоску)
   * Параметр hideBottomDivider скрывает нижнюю серую полоску только у последнего AccordionItem, заданного внутри Accordion
   */
  hideBottomDivider?: boolean;
}

export const Accordion: FC<AccordionProps> = ({
  children,
  dimension = 'l',
  iconPosition = 'right',
  hideTopDivider = false,
  hideBottomDivider = false,
  onKeyDown,
  ...props
}) => {
  const accordionRef = useRef<HTMLDivElement | null>(null);
  const handleKeyDown = useCallback((e: KeyboardEvent<HTMLDivElement>) => {
    let focusedOption: any = ((accordionRef.current && accordionRef.current.ownerDocument) || document).activeElement;
    // если фокус находится на AccordionItem Title, то можно переключаться между accordion items с помощью ArrowDown, ArrowUp, Home, End
    if (focusedOption.id.indexOf('accordion_title') > -1 && focusedOption.tagName.toLowerCase() === 'div') {
      do {
        focusedOption = focusedOption?.parentNode;
      } while (focusedOption.parentNode !== accordionRef.current);

      const code = keyboardKey.getCode(e);

      if (code === keyboardKey.ArrowDown) {
        moveFocus(accordionRef.current, focusedOption, nextItem);
        e.preventDefault();
      } else if (code === keyboardKey.ArrowUp) {
        moveFocus(accordionRef.current, focusedOption, previousItem);
        e.preventDefault();
      } else if (code === keyboardKey.Home) {
        e.preventDefault();
        moveFocus(accordionRef.current, null, nextItem);
      } else if (code === keyboardKey.End) {
        e.preventDefault();
        moveFocus(accordionRef.current, null, previousItem);
      }
    }
    onKeyDown?.(e);
  }, []);
  return (
    <AccordionWrapper
      ref={accordionRef}
      data-dimension={dimension}
      data-icon={iconPosition}
      $hideTopDivider={hideTopDivider}
      $hideBottomDivider={hideBottomDivider}
      $dimension={dimension}
      onKeyDown={handleKeyDown}
      {...props}
    >
      {children}
    </AccordionWrapper>
  );
};

Accordion.displayName = 'Accordion';
