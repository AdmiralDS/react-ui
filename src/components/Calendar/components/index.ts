export * from './Years';
export * from './Button';
export * from './Day';
export * from './DayNames';
export * from './Month';
export * from './Months';
export * from './Panel';
export * from './Week';
