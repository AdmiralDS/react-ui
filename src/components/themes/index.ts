export * from './light';
export * from './dark';
export type { Theme } from './common';
