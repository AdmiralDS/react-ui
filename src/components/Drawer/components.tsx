import type { FC, HTMLAttributes } from 'react';
import { createContext, useContext, useLayoutEffect, useRef } from 'react';
import styled from 'styled-components';

import { typography } from '#src/components/Typography';
import { debounce } from '#src/components/common/utils/debounce';
import { resizePaddings } from '../common/dom/resizePaddings';

const Title = styled.h5<{ $mobile: boolean; $displayCloseIcon: boolean }>`
  ${({ $mobile }) => ($mobile ? typography['Header/H6'] : typography['Header/H5'])};
  color: var(--admiral-color-Neutral_Neutral90, ${(p) => p.theme.color['Neutral/Neutral 90']});
  margin: 0;
  padding: ${({ $mobile, $displayCloseIcon }) => {
    if ($mobile) {
      return $displayCloseIcon ? '0 48px 8px 16px' : '0 16px 8px';
    }
    return $displayCloseIcon ? '0 56px 8px 24px' : '0 24px 8px';
  }};
`;

const Content = styled.div<{ $mobile: boolean }>`
  overflow-y: auto;
  outline: none;
  scrollbar-gutter: stable;
  padding-block: 8px;
  padding-inline: ${({ $mobile }) => `${$mobile ? 16 : 24}px`};
  height: 100%;
`;

const ButtonPanel = styled.div<{ $mobile: boolean }>`
  display: flex;
  flex-direction: ${({ $mobile }) => ($mobile ? 'column-reverse' : 'row-reverse')};
  padding: ${({ $mobile }) => ($mobile ? '16px 16px 0' : '16px 24px 0')};

  & > button {
    margin: ${({ $mobile }) => ($mobile ? '0 0 16px 0' : '0 16px 0 0')};
    ${({ $mobile }) => $mobile && 'width: 100%;'}
  }

  & > button:first-child {
    margin: 0;
  }
`;

export const DrawerContext = createContext({ mobile: false, displayCloseIcon: true } as {
  mobile: boolean;
  displayCloseIcon: boolean;
});

export const DrawerTitle: FC<HTMLAttributes<HTMLHeadingElement>> = ({ children, ...props }) => {
  const { mobile, displayCloseIcon } = useContext(DrawerContext);
  const asProp = mobile ? 'h6' : 'h5';
  return (
    <Title $mobile={mobile} $displayCloseIcon={displayCloseIcon} as={asProp} {...props}>
      {children}
    </Title>
  );
};

export const DrawerContent: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const contentRef = useRef<HTMLDivElement | null>(null);
  const mobile = useContext(DrawerContext).mobile;

  useLayoutEffect(() => {
    const node = contentRef.current;
    if (node) {
      resizePaddings(node);
      const resizeObserver = new ResizeObserver(debounce(() => resizePaddings(node), 250));
      resizeObserver.observe(node);
      return () => {
        resizeObserver.unobserve(node);
      };
    }
  }, []);

  return (
    <Content tabIndex={-1} ref={contentRef} $mobile={mobile} {...props}>
      {children}
    </Content>
  );
};

export const DrawerButtonPanel: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const mobile = useContext(DrawerContext).mobile;
  return (
    <ButtonPanel $mobile={mobile} {...props}>
      {children}
    </ButtonPanel>
  );
};
