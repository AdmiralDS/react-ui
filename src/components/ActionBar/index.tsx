import { Fragment, useLayoutEffect, useMemo, useRef, useState } from 'react';
import styled from 'styled-components';

import type { ActionBarDimension, ActionBarItemProps, ActionBarProps } from '#src/components/ActionBar/types';
import {
  ACTION_BAR_ITEM_HEIGHT_L,
  ACTION_BAR_ITEM_HEIGHT_M,
  ACTION_BAR_ITEM_HEIGHT_S,
  ACTION_BAR_ITEM_HEIGHT_XL,
  DIVIDER_SIZE_M,
  DIVIDER_SIZE_S,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_L,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_M,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_S,
  FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_XL,
} from '#src/components/ActionBar/constants';
import { OverflowMenu } from '#src/components/OverflowMenu';
import { ActionBarDivider } from '#src/components/ActionBar/ActionBarItems';
import type { MenuModelItemProps } from '#src/components/Menu/MenuItem';
import { mediumGroupBorderRadius } from '#src/components/themes';

export * from './types';
export * from './ActionBarItems';

const getActionBarHeight = (dimension: ActionBarDimension) => {
  switch (dimension) {
    case 's':
      return ACTION_BAR_ITEM_HEIGHT_S;
    case 'm':
      return ACTION_BAR_ITEM_HEIGHT_M;
    case 'l':
      return ACTION_BAR_ITEM_HEIGHT_L;
    case 'xl':
    default:
      return ACTION_BAR_ITEM_HEIGHT_XL;
  }
};

const Wrapper = styled.div<{ $dimension: ActionBarDimension }>`
  box-sizing: border-box;
  height: ${({ $dimension }) => getActionBarHeight($dimension)}px;

  display: flex;
  align-items: center;
`;
const ActionBarOverflowMenuItem = styled.div<{ $dimension: ActionBarDimension }>`
  flex-shrink: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: ${({ $dimension }) => getActionBarHeight($dimension)}px;
  height: ${({ $dimension }) => getActionBarHeight($dimension)}px;
`;
const getFocusBorderOffset = ($dimension: ActionBarDimension) => {
  switch ($dimension) {
    case 's':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_S;
    case 'm':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_M;
    case 'l':
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_L;
    case 'xl':
    default:
      return FOCUS_ACTION_BAR_OVERFLOW_MENU_BORDER_OFFSET_XL;
  }
};

function getFocusBorderRadius(componentBorderRadius: string) {
  switch (componentBorderRadius) {
    case '0':
      return '0';
    case '2px':
      return '6px';
    case '6px':
      return '10px';
    case '8px':
      return '12px';
    case '10px':
      return '14px';
    case '4px':
    default:
      return '8px';
  }
}

const StyledOverflowMenu = styled(OverflowMenu)<{ $dimension: ActionBarDimension }>`
  &:focus-visible:not(:disabled) {
    outline: none;

    &:before {
      border: 2px solid var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']});
      border-radius: ${(p) => getFocusBorderRadius(mediumGroupBorderRadius(p.theme.shape))};
      content: '';
      display: block;
      position: absolute;
      width: ${({ $dimension }) => getActionBarHeight($dimension) + 2 * 2}px;
      height: ${({ $dimension }) => getActionBarHeight($dimension) + 2 * 2}px;
      top: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      left: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      right: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      bottom: -${(p) => getFocusBorderOffset(p.$dimension)}px;
      pointer-events: none;
    }
  }
`;

export const ActionBar = ({
  dimension = 'xl',
  renderActionBarItem,
  renderDropMenuItem,
  items,
  itemIsDisabled,
  ...props
}: ActionBarProps) => {
  const renderItem = ({ itemId, withDivider }: ActionBarItemProps) => {
    return (
      <Fragment key={itemId}>
        {renderActionBarItem(itemId)}
        {withDivider && <ActionBarDivider key={`divider${itemId}`} dimension={dimension} />}
      </Fragment>
    );
  };

  const [containerWidth, setContainerWidth] = useState(0);
  const containerRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    if (containerRef.current) {
      const resizeObserver = new ResizeObserver((entries) => {
        entries.forEach((entry) => setContainerWidth(entry.contentRect.width || 0));
      });
      resizeObserver.observe(containerRef.current);
      return () => {
        resizeObserver.disconnect();
      };
    }
  }, []);
  const [visibleItems, setVisibleItems] = useState<ActionBarItemProps[]>([]);
  const [hiddenItems, setHiddenItems] = useState<ActionBarItemProps[]>([]);

  useLayoutEffect(() => {
    if (containerRef.current) {
      const buttonWidth = getActionBarHeight(dimension);
      const dividerWidth = dimension === 's' ? DIVIDER_SIZE_S : DIVIDER_SIZE_M;

      const totalItemsWidth = items.reduce((acc, item) => {
        return acc + buttonWidth + (item.withDivider ? dividerWidth : 0);
      }, 0);
      const overflow = totalItemsWidth > containerWidth;

      let availableWidth = containerWidth - (overflow ? buttonWidth : 0);

      const newVisibleItems: ActionBarItemProps[] = [];
      const newHiddenItems: ActionBarItemProps[] = [];
      const addToVisible = (item: ActionBarItemProps) => newVisibleItems.push(item);
      const addToHidden = (item: ActionBarItemProps) => newHiddenItems.push(item);

      items.forEach((item) => {
        const itemWidth = buttonWidth + (item.withDivider ? dividerWidth : 0);
        if (availableWidth >= itemWidth) {
          addToVisible(item);
          availableWidth -= itemWidth;
        } else {
          addToHidden(item);
          availableWidth = 0;
        }
      });

      setVisibleItems(newVisibleItems);
      setHiddenItems(newHiddenItems);
    }
  }, [items, containerWidth, dimension]);
  const overflowMenuItems: MenuModelItemProps[] = useMemo(() => {
    if (hiddenItems.length === 0) return [];
    return hiddenItems.map(({ itemId }) => {
      return {
        id: itemId,
        render: renderDropMenuItem(itemId),
        disabled: itemIsDisabled(itemId),
      };
    });
  }, [hiddenItems, dimension]);

  return (
    <Wrapper {...props} ref={containerRef} $dimension={dimension}>
      {visibleItems.map((item) => renderItem(item))}
      {hiddenItems.length > 0 && (
        <ActionBarOverflowMenuItem $dimension={dimension}>
          <StyledOverflowMenu
            $dimension={dimension}
            items={overflowMenuItems}
            dimension={dimension === 'xl' ? 'l' : dimension}
          />
        </ActionBarOverflowMenuItem>
      )}
    </Wrapper>
  );
};
