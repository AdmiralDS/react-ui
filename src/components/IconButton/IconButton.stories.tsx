import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled, { ThemeProvider } from 'styled-components';
import { IconButton } from './index';
import { ReactComponent as StarSolid } from '@admiral-ds/icons/build/system/StarSolid.svg';
import { T } from '#src/components/T';
import { Theme } from '#src/components/themes';

const WrapperIconButton = styled.div`
  display: flex;
  flex-wrap: wrap;
  > * {
    margin: 8px;
  }
  justify-content: space-between;
  align-items: center;
`;

export default {
  title: 'Admiral-2.1/IconButton',
  decorators: [withDesign],
  component: IconButton,
  parameters: {
    docs: {
      source: {
        code: null,
      },
    },
    layout: 'centered',
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A17787',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A18110',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A18298',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A18322',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A18345',
      },
    ],
  },
  argTypes: {
    dimension: {
      options: ['xl', 'l', 'm', 's'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    loading: {
      control: { type: 'boolean' },
    },
    themeBorderKind: {
      control: {
        type: 'radio',
        options: ['Border radius 0', 'Border radius 2', 'Border radius 4', 'Border radius 8'],
      },
    },
    skeleton: {
      control: { type: 'boolean' },
    },
  },
} as ComponentMeta<typeof IconButton>;

const IconButtonDemo: ComponentStory<typeof IconButton> = (args) => {
  function swapBorder(theme: Theme): Theme {
    theme.shape.borderRadiusKind = (args as any).themeBorderKind || theme.shape.borderRadiusKind;
    return theme;
  }

  return (
    <ThemeProvider theme={swapBorder}>
      <WrapperIconButton>
        <div>
          <T font="Body/Body 1 Long" as="div">
            Dimension - xl
          </T>
          <IconButton {...args} dimension="xl">
            <StarSolid />
          </IconButton>
        </div>
        <div>
          <T font="Body/Body 1 Long" as="div">
            Dimension - l
          </T>
          <IconButton {...args} dimension="l">
            <StarSolid />
          </IconButton>
        </div>
        <div>
          <T font="Body/Body 1 Long" as="div">
            Dimension - m
          </T>
          <IconButton {...args} dimension="m">
            <StarSolid />
          </IconButton>
        </div>
        <div>
          <T font="Body/Body 1 Long" as="div">
            Dimension - s
          </T>
          <IconButton {...args} dimension="s">
            <StarSolid />
          </IconButton>
        </div>
      </WrapperIconButton>
    </ThemeProvider>
  );
};

const IconButtonStateDemo: ComponentStory<typeof IconButton> = () => (
  <>
    <WrapperIconButton>
      <div>
        <T font="Body/Body 1 Long" as="div">
          Loading
        </T>
        <IconButton loading dimension="xl" />
      </div>
      <div>
        <T font="Body/Body 1 Long" as="div">
          Disabled
        </T>
        <IconButton disabled dimension="xl">
          <StarSolid />
        </IconButton>
      </div>
      {/*<div>*/}
      {/*  <T font="Body/Body 1 Long" as="div">*/}
      {/*    Skeleton*/}
      {/*  </T>*/}
      {/*  <IconButton skeleton dimension="xl" />*/}
      {/*</div>*/}
    </WrapperIconButton>
  </>
);

export const IconButtons = IconButtonDemo.bind({});
IconButtons.args = {};
IconButtons.storyName = 'IconButton. Размеры';
export const IconButtonState = IconButtonStateDemo.bind({});
IconButtonState.args = {};
IconButtonState.storyName = 'IconButton. Состояния';
