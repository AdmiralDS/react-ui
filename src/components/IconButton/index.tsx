import type { ButtonHTMLAttributes } from 'react';
import { forwardRef } from 'react';
import styled from 'styled-components';

import { SpinnerIcon } from '#src/components/Spinner/SpinnerIcon';
import { mediumGroupBorderRadius } from '#src/components/themes/borderRadius';
import { skeletonAnimationMixin } from '#src/components/skeleton/animation';

type Dimension = 'xl' | 'l' | 'm' | 's';
type IconButtonAppearance = 'primary' | 'secondary';

export interface IconButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Внешний вид кнопки */
  appearance?: IconButtonAppearance;
  /** Размер кнопки */
  dimension?: Dimension;
  /** Отключение кнопки */
  disabled?: boolean;
  /** Состояние loading */
  loading?: boolean;
  /** Состояние skeleton */
  skeleton?: boolean;
}

const StyledButton = styled.button<{
  $skeleton?: boolean;
  $loading?: boolean;
  $appearance?: IconButtonAppearance;
}>`
  box-sizing: border-box;
  display: inline-block;
  border: none;
  border-radius: ${(p) =>
    p.$skeleton ? 0 : `var(--admiral-border-radius-Medium, ${mediumGroupBorderRadius(p.theme.shape)})`};
  background: ${({ $skeleton }) => ($skeleton ? 'red' : 'transparent')};
  -webkit-tap-highlight-color: transparent;
  appearance: none;
  vertical-align: middle;

  .icon-button-group[data-dimension='xl'] &,
  &[data-dimension='xl'] {
    padding: 0;
    height: 56px;
    width: 56px;
  }

  .icon-button-group[data-dimension='l'] &,
  &[data-dimension='l'] {
    height: 48px;
    width: 48px;
  }

  .icon-button-group[data-dimension='m'] &,
  &[data-dimension='m'] {
    height: 40px;
    width: 40px;
  }

  .icon-button-group[data-dimension='s'] &,
  &[data-dimension='s'] {
    height: 32px;
    width: 32px;
  }

  &:hover {
    cursor: ${({ $skeleton }) => ($skeleton ? 'default' : 'pointer')};
    background: var(--admiral-color-Opacity_Hover, ${(p) => p.theme.color['Opacity/Hover']});
    & *[fill^='#'] {
      fill: ${(p) =>
        p.$appearance === 'primary'
          ? `var(--admiral-color-Primary_Primary60Main, ${p.theme.color['Primary/Primary 60 Main']})`
          : `var(--admiral-color-Neutral_Neutral50, ${p.theme.color['Neutral/Neutral 50']})`};
    }
  }
  &:active {
    background: var(--admiral-color-Opacity_Press, ${(p) => p.theme.color['Opacity/Press']});
    & *[fill^='#'] {
      fill: ${(p) =>
        p.$appearance === 'primary'
          ? `var(--admiral-color-Primary_Primary60Main, ${p.theme.color['Primary/Primary 60 Main']})`
          : `var(--admiral-color-Neutral_Neutral50, ${p.theme.color['Neutral/Neutral 50']})`};
    }
  }

  &:disabled {
    cursor: ${({ $loading }) => ($loading ? 'default' : 'not-allowed')};
    background-color: transparent;
    & *[fill^='#'] {
      fill: var(--admiral-color-Neutral_Neutral30, ${(p) => p.theme.color['Neutral/Neutral 30']});
    }
  }

  &:focus-visible {
    outline-offset: 2px;
    outline: var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']}) solid 2px;
  }

  ${({ $skeleton }) => $skeleton && skeletonAnimationMixin};
`;

const IconButtonContent = styled.span<{ $dimension?: Dimension; $appearance?: IconButtonAppearance }>`
  vertical-align: top;
  display: inline-flex;
  flex-direction: row;
  overflow: hidden;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: center;
  width: ${({ $dimension }) => ($dimension === 's' ? 20 : 24)}px;
  height: ${({ $dimension }) => ($dimension === 's' ? 20 : 24)}px;

  .icon-button-group[data-dimension='m'] &&,
  .icon-button-group[data-dimension='l'] &&,
  .icon-button-group[data-dimension='xl'] && {
    width: 24px;
    height: 24px;
  }
  .icon-button-group[data-dimension='s'] && {
    width: 20px;
    height: 20px;
  }

  > * {
    display: inline-block;
    flex: 0 1 auto;
    white-space: nowrap;
  }

  > *:not(:first-child) {
    margin-left: 8px;
  }

  & *[fill^='#'] {
    fill: ${(p) =>
      p.$appearance === 'primary'
        ? `var(--admiral-color-Primary_Primary60Main, ${p.theme.color['Primary/Primary 60 Main']})`
        : `var(--admiral-color-Neutral_Neutral50, ${p.theme.color['Neutral/Neutral 50']})`};
  }

  & > svg {
    width: 100%;
    height: 100%;
  }
`;

const SpinnerContainer = styled(IconButtonContent)`
  container-type: inline-size;
`;

export const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  (
    {
      appearance = 'secondary',
      dimension = 'xl',
      type = 'button',
      loading = false,
      skeleton = false,
      disabled = false,
      children,
      ...props
    }: IconButtonProps,
    ref,
  ) => {
    const disabledOptions = loading || disabled;
    const renderContent = () => {
      if (loading) {
        return (
          <SpinnerContainer $dimension={dimension}>
            <SpinnerIcon />
          </SpinnerContainer>
        );
      }
      if (skeleton) {
        return <IconButtonContent $dimension={dimension} />;
      }
      return (
        <IconButtonContent $dimension={dimension} $appearance={appearance}>
          {children}
        </IconButtonContent>
      );
    };

    return (
      <StyledButton
        ref={ref}
        data-dimension={dimension}
        disabled={disabledOptions}
        type={type}
        $skeleton={skeleton}
        $loading={loading}
        $appearance={appearance}
        {...props}
      >
        {renderContent()}
      </StyledButton>
    );
  },
);

IconButton.displayName = 'IconButton';
