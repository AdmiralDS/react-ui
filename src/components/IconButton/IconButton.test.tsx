import { ThemeProvider } from 'styled-components';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { LIGHT_THEME } from '#src/components/themes';
import { IconButton } from '#src/components/IconButton';

describe('IconButton', () => {
  test('should render component', () => {
    const { container } = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <IconButton />
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should focus on component when user on press key tab', async () => {
    const user = userEvent.setup();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <IconButton />
      </ThemeProvider>,
    );
    expect(document.body).toHaveFocus();
    await user.tab();
    expect(screen.getByRole('button')).toHaveFocus();
  });

  test('should not focus component when user on press key tab', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <IconButton disabled />
      </ThemeProvider>,
    );
    expect(document.body).toHaveFocus();
    userEvent.tab();
    expect(document.body).toHaveFocus();
  });

  test('should call onClick when user clicks on component', () => {
    const handleClick = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <IconButton onClick={handleClick} />
      </ThemeProvider>,
    );
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
