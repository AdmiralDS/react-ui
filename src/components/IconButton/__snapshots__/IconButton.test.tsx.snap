// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should render component 1`] = `
.c0 {
  box-sizing: border-box;
  display: inline-block;
  border: none;
  border-radius: var(--admiral-border-radius-Medium, 4px);
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  appearance: none;
  vertical-align: middle;
}

.icon-button-group[data-dimension='xl'] .c0,
.c0[data-dimension='xl'] {
  padding: 0;
  height: 56px;
  width: 56px;
}

.icon-button-group[data-dimension='l'] .c0,
.c0[data-dimension='l'] {
  height: 48px;
  width: 48px;
}

.icon-button-group[data-dimension='m'] .c0,
.c0[data-dimension='m'] {
  height: 40px;
  width: 40px;
}

.icon-button-group[data-dimension='s'] .c0,
.c0[data-dimension='s'] {
  height: 32px;
  width: 32px;
}

.c0:hover {
  cursor: pointer;
  background: var(--admiral-color-Opacity_Hover, #0000000D);
}

.c0:hover *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

.c0:active {
  background: var(--admiral-color-Opacity_Press, #0000001F);
}

.c0:active *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: transparent;
}

.c0:disabled *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

.c0:focus-visible {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

.c1 {
  vertical-align: top;
  display: inline-flex;
  flex-direction: row;
  overflow: hidden;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: center;
  width: 24px;
  height: 24px;
}

.icon-button-group[data-dimension='m'] .c1.c1,
.icon-button-group[data-dimension='l'] .c1.c1,
.icon-button-group[data-dimension='xl'] .c1.c1 {
  width: 24px;
  height: 24px;
}

.icon-button-group[data-dimension='s'] .c1.c1 {
  width: 20px;
  height: 20px;
}

.c1 >* {
  display: inline-block;
  flex: 0 1 auto;
  white-space: nowrap;
}

.c1 >*:not(:first-child) {
  margin-left: 8px;
}

.c1 *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

.c1>svg {
  width: 100%;
  height: 100%;
}

<div>
  <button
    class="c0"
    data-dimension="xl"
    type="button"
  >
    <span
      class="c1"
    />
  </button>
</div>
`;
